include/El/optimization/bfgs.hpp

```diff
@@ -0,0 +1,423 @@
+ * (a,b,c,d) = (invert {{x^3, x^2, x, 1},{t^3, t^2, t, 1}, {3x^2, 2x, 1, 0}, {3t^2, 2t, 1, 0}})*{{fX},{fT},{gX},{gT}}
+ */
+template< typename T, typename Function>
+inline T MinCubicInterpolate(T x, T fX, T t, T fT, T gX, T gT, const Function& phi){
+     T xSquared = x*x;
+     T tSquared = t*t;
+     T invDet = (T(1)/((t-x)*(t-x)));
+     T a = ((T(2)*fX - T(2)*fT)/(t-x) + gX + gT);
+     T b = (((t+x)/(t-x))*(T(-3)*fX + 3*gT) + (T(-2)*t-x)*gX  + (-t-T(2)*x)*gT);
+     T c = ((T(6)*t*x/(t-x))*(fX-fT) + (tSquared + 2*x*t)*gX  + (xSquared + 2*t*x)*gT);
+     T discriminant = sqrt(b*b - 3*a*c)*(1/(t-x));
+     b*=invDet;
+     a*=invDet;
+     T root1 = (-b + discriminant)/(T(3)*a);
+     T root2 = (-b - discriminant)/(T(3)*a);
+     T f1 = phi(root1);
+     T f2 = phi(root2);
+     if( f1 <= f2){ return root1; }
+     return root2;
+ }
+
+/**
+ * This technique is borrowed from the paper:
+ * Line Search Algorithms with Guaranteed Sufficient Decrease
+ * By JORGE J. MORE and DAVID J. THUENTE
+ * Section 4, Trial Value Selection
+ * https://pdfs.semanticscholar.org/d258/1519560dd59e5b92599f9711aa1ab249ff86.pdf
+
+ * @param alphaLow
+ * @param alphaHigh
+ * @param alpha
+ * @param phiLow
+ * @param phiHigh
+ * @param phiAlpha
+ * @param dPhiLow
+ * @param dphiAlpha
+ * @return
+ */
+template< typename Function, typename T>
+inline T Interpolate(T alphaLow, T alphaHigh, T alpha, T phiLow, T phiHigh, T phiAlpha,
+              T dPhiLow, T dPhiHigh, T dPhiAlpha, const Function& phi){
+
+            T alphaC = MinCubicInterpolate(alphaLow, phiLow, alpha, phiAlpha, dPhiLow, dPhiAlpha, phi);
+            if( phiAlpha > phiLow){
+                T alphaQ = MinQuadraticInterpolate(alphaLow, phiLow, alpha, phiAlpha, dPhiLow);
+                if( Abs(alphaC - alphaLow) < Abs(alphaQ - alphaLow)){ return alphaC; }
+                return (alphaQ+alphaC)/T(2);
+            }
+            T alphaS = MinQuadraticInterpolateTwoDerivatives(alphaLow, dPhiLow, alpha, dPhiAlpha);
+            if(  phiAlpha < phiLow and dPhiAlpha*dPhiLow < 0){
+                if( Abs(alphaC - alpha) >= Abs(alphaS - alpha)){ return alphaC; }
+                return alphaS;
+            }
+            if(  phiAlpha < phiLow and dPhiAlpha*dPhiLow > 0 and Abs(dPhiAlpha) <= Abs(dPhiLow)){
+                if( Abs(alphaC - alpha) < Abs(alphaS - alpha)){ return alphaC; }
+                return alphaS;
+            }
+            //Else: phiAlpha <= phiLow and dPhiAlpha*dPhiLow > 0 and Abs(dPhiAlpha) > Abs(dPhiLow))
+            return  MinCubicInterpolate(alphaHigh, phiHigh, alpha, phiAlpha, dPhiHigh, dPhiAlpha, phi);
+}
+
+/**
+ * Preconditions:
+ * -------------
+ * 1. The interval bounded by alphaLow and alphaHigh contains a value alpha satisfying the strong wolfe conditions.
+ * 2. alphaLow satisfies the sufficient decrease condition.
+ * 3. f(x + alphaLow*p)*(alphaHigh - alphaLow) < 0
+ *
+ * Note: it is _not_ required that alphaLow < alphaHigh.
+ *
+ * @param f
```

Comment:
I would strongly recommend documenting what each routine is meant to accomplish before documenting the parameters.